c7n/resources/iam.py

```diff
@@ -435,57 +435,53 @@ def match(self, info):
-                UserName=resource['UserName'])['AttachedPolicies']
+    def user_policies(self, resource):
+        if 'c7n.AttachedPolicies' in resource:
+            return
+        client = local_session(self.manager.session_factory).client('iam')
+        resource['c7n.AttachedPolicies'] = client.list_attached_user_policies(
+            UserName=resource['UserName'])['AttachedPolicies']
 
+    def process(self, resources, event=None):
         with self.executor_factory(max_workers=2) as w:
-            query_resources = [
-                r for r in resources if 'AttachedPolicies' not in r]
             self.log.debug(
-                "Querying %d users policies" % len(query_resources))
-            list(w.map(_user_policies, query_resources))
+                "Querying %d users policies" % len(resources))
+            list(w.map(self.user_policies, resources))
 
-        matched = []
-        for r in resources:
-            for p in r['AttachedPolicies']:
-                if self.match(p):
-                    matched.append(r)
-                    break
-        return matched
+        if self.data.get('attached', True):
+            return [r for r in resources if r['c7n.AttachedPolicies']]
+        else:
+            return [r for r in resources if not r['c7n.AttachedPolicies']]
 
 
 @User.filter_registry.register('access-key')
-class UserAccessKey(ValueFilter):
+class UserAccessKey(Filter):
 
-    schema = type_schema('access-key', rinherit=ValueFilter.schema)
-
-    def process(self, resources, event=None):
+    schema = type_schema(
+        'access-key',
+        present={'type': 'boolean'})
 
-        def _user_keys(resource):
-            client = local_session(self.manager.session_factory).client('iam')
-            resource['AccessKeys'] = client.list_access_keys(
-                UserName=resource['UserName'])['AccessKeyMetadata']
+    def user_keys(self, resource):
+        if 'c7n.AccessKeys' in resource:
+            return
+        client = local_session(self.manager.session_factory).client('iam')
+        resource['c7n.AccessKeys'] = client.list_access_keys(
+            UserName=resource['UserName'])['AccessKeyMetadata']
 
+    def process(self, resources, event=None):
         with self.executor_factory(max_workers=2) as w:
-            query_resources = [
-                r for r in resources if 'AccessKeys' not in r]
             self.log.debug(
-                "Querying %d users' api keys" % len(query_resources))
-            list(w.map(_user_keys, query_resources))
+                "Querying %d users' api keys" % len(resources))
+            list(w.map(self.user_keys, resources))
 
-        matched = []
-        for r in resources:
-            for p in r['AccessKeys']:
-                if self.match(p):
-                    matched.append(r)
-                    break
-        return matched
+        if self.data.get('present', True):
```

Comment:
fwiw, these filters weren't intended as a boolean existence test, but a match on key attributes like CreationDate, which the previous impl tried to do via match().